op-proposer/proposer/driver.go

```diff
@@ -478,26 +497,37 @@ func (l *L2OutputSubmitter) loopL2OO(ctx context.Context) {
 	}
 }
 
+// The loopDGF proposes a new output every proposal interval. It does _not_ query
+// the DGF for when to next propose, as the DGF doesn't have the concept of a
+// proposal interval, like in the L2OO case. For this reason, it has to keep track
+// of the interval itself, for which it uses an internal ticker.
 func (l *L2OutputSubmitter) loopDGF(ctx context.Context) {
+	defer l.Log.Info("loopDGF returning")
 	ticker := time.NewTicker(l.Cfg.ProposalInterval)
 	defer ticker.Stop()
 	for {
 		select {
 		case <-ticker.C:
-			blockNumber, err := retry.Do(ctx, 10, &retry.FixedStrategy{Dur: l.Cfg.OutputRetryInterval}, func() (*big.Int, error) {
-				return l.FetchCurrentBlockNumber(ctx)
-			})
-			if err != nil {
-				break
+			// prioritize quit signal
+			select {
+			case <-l.done:
+				return
+			default:
 			}
 
-			output, shouldPropose, err := retry.Do2(ctx, 10, &retry.FixedStrategy{Dur: l.Cfg.OutputRetryInterval}, func() (*eth.OutputResponse, bool, error) {
-				ctx, cancel := context.WithTimeout(ctx, 3*time.Second)
-				defer cancel()
-				return l.FetchOutput(ctx, blockNumber)
-			})
-			if err != nil || !shouldPropose {
-				break
+			var (
+				output *eth.OutputResponse
+				err    error
+			)
+			// A note on retrying: because the proposal interval is usually much
+			// larger than the interval at which to retry proposing on a failed attempt,
+			// we want to keep retrying getting the output proposal until we succeed.
+			for output == nil || err != nil {
+				output, err = l.FetchDGFOutput(ctx)
+				if err != nil {
```

Comment:
Should this be `if output == nil || err != nil` to match the `for` loop condition above? Or do we only care about the `err` value inside this loop?

Reply:
The `output == nil` condition is for the first iteration, when `err == nil`. When we check the error, we already set `output, err`, so we only need to check for the error.